numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -1627,12 +1627,31 @@ _array_fromobject_generic(
                 goto finish;
             }
         }
-        /* One more chance */
+        /* One more chance for faster exit if user specified the dtype. */
```

Comment:
It seems like the whole conditional (starting at 1612) might be appropriately replaced with a call to `array_astype()`, except that it looks like there are some default behaviors that are different (e.g. forcecopy).

In particular, the use of `PyArray_DescrConverter -> PyArray_AdaptDescriptorToArray` in `methods.c` `array_astype()` has slightly different behavior than the `PyArray_DescrConverter2` -> `PyArray_DescrConverter` here.

It seems like a separate issue, but maybe any associated behavior changes are something that should be considered before release notes are finalized?

Reply:
Its a good point that it may be nice to consolidate the fast-path with `astype`.  But I think its OK to not do it for now.